### PR TITLE
Print `methods(f)` with color like stack traces

### LINKED ISSUE #30110: Colorize output of `methods` in REPL
The output of `methods` can be hard to read in the REPL
It would be good to use some colour (/bold) 
to seperate the signature from the line-info.

You can see that in the stacktrace at the top of the image below.
Contrast how easy to read it is, vs the methods list at the bottom.

image

### LINKED ISSUE #36129: Can we prettify how methods are printed as in #36026?
The current output of `ClearStacktrace.jl` is pretty good. I just wonder if the same technique discussed in #36026 can be used on how methods are printed in `methods` and `methodswith`? Considering the different result types of them, maybe we need to redefine how methods are shown?

```julia
julia> methodswith(Int) |> typeof
Array{Method,1}

julia> methods(Int) |> typeof
Base.MethodList
```

### LINKED ISSUE #40913: Colourising `Method` representation for easier glancing
Hello!

I love being able to see all the various implement methods, or methods using a certain type in Julia.
However, I didn't find it easy to find what I was looking for at a glance. Particularly for functions with many (tens-hundreds) of implementations. So, I've tried adding a bit of colour. I've found it really helpful, and think it would be good if something like it were the default.

Happy to make a PR, or just see my work used.

### Example

Current output for `methods(sum)`

image

My output for `methods(sum)`

image

### Code

```julia
function Base.show(io::IO, m::Method)
    tv, decls, file, line = Base.arg_decl_parts(m)
    sig = Base.unwrap_unionall(m.sig)
    if sig === Tuple
        # Builtin
        print(io, m.name, "(...) in ", m.module)
        return
    end
    printstyled(io, decls[1][2]; color=:blue)
    print(io, "(")
    for (d, i) in zip(decls[2:end], 2:length(decls))
        if isempty(d[2])
            printstyled(io, d[1]; color=:magenta)
        else
            printstyled(io, d[1]; color=:magenta)
            typespec = split(d[2], "{"; limit=2)
            if length(typespec) == 1
                printstyled(io, "::", d[2]; color=:yellow)
            else
                printstyled(io, "::", typespec[1]; color=:yellow)
                printstyled(io, "{", typespec[2]; color=247)
            end
        end
        if i < length(decls)
            printstyled(io, ", ")
        end
    end
    kwargs = Base.kwarg_decl(m)
    if !isempty(kwargs)
        printstyled(io, "; "; bold=true)
        for (k, i) in zip(kwargs, 1:length(kwargs))
            printstyled(io, Base.sym_to_string(k); color=:magenta)
            if i < length(kwargs)
                printstyled(io, ", ")
            end
        end
    end
    print(io, ")")
    Base.show_method_params(io, tv)
    printstyled(io, " in "; color=:light_black)
    printstyled(io, m.module; color=:light_red)
    if line > 0
        file, line = Base.updated_methodloc(m)
        printstyled(io, " at "; color=:light_black)
        printstyled(io, file; color=:green)
        printstyled(io, ":", line)
    end
end

function Base.show_method_list_header(io::IO, ms::Base.MethodList, namefmt::Function)
    mt = ms.mt
    name = mt.name
    hasname = isdefined(mt.module, name) &&
              typeof(getfield(mt.module, name)) <: Function
    n = length(ms)
    if mt.module === Core && n == 0 && mt.defs === nothing && mt.cache !== nothing
        # try to detect Builtin
        print(io, "# built-in function; no methods")
    else
        printstyled(io, "# "; color=:light_black)
        printstyled(io, n; color=:cyan, bold=true)
        print(io, " ", n==1 ? "method" : "methods")
        sname = string(name)
        if hasname
            what = startswith(sname, '@') ? "macro" : "generic function"
            print(io, " for ", what, " ")
            printstyled(io, sname; color=:blue)
        elseif '#' in sname
            print(io, " for anonymous function ", namefmt(sname))
        elseif mt === Base._TYPE_NAME.mt
            print(io, " for type constructor")
        end
        print(io, ":")
    end
end
```

### PR DESCRIPTION
**Pull request by mcabbott**
Should we re-use something like the new error printing for method lists? 

Perhaps the module `which(f)` should be mostly omitted, to highlight only other-module methods?

Screenshot 2021-03-29 at 00 37 12

(Edit -- deleted an example. The one above does not yet have types in bold, added below.)

Closes #40913 Closes #30110 Closes #36129

### CODE REVIEW

**Comment by thchr** [comment]
This is no doubt subjective, but when I'm calling `methods` what I'm most interested in disambiguating is the different type signatures. This is somewhat different from what one is interested in emphasizing for stack traces (e.g. different modules and path locations) and what is most clearly highlighted with this coloring/highlighting.

**Comment by fingolfin** [comment]
I really like this. However, I feel that the colored package names in your examples stand out most -- they are the first thing I see/read when looking at those images. I feel that's reverse from how it should be. Perhaps one can experiment a bit more with this. Either by deemphasizing the package names, or by emphasizing the method argument types more. Besides colors, one could also experiment with e.g. using bold face -- e.g. how does it look if the argument types are printed in bold?

**Comment by mcabbott** [comment]
More subtle colours would be better, but hard to make things precise. Bold seems to vary from doing nothing to feeling like shouting, here's a very crude attempt:

Screenshot 2021-03-30 at 11 44 13

With more fiddling the colour could be moved to apply to the types owned by that package, not the module name. Or (simpler) to highlight the package name in the path, rather than alone.

**Comment by fingolfin** [comment]
To me your last screenshot is by far the best in terms if being able to quickly pick out which methods are there and in particular finding any that I care about

**Comment by StefanKarpinski** [comment]
Really like these! Seems ok if the coloring and emphasis of stack traces and methods lists are a bit different, but ideal to share the common printing logic aside from those details.

**Comment by fingolfin** [comment]
Nice!

Any idea why in the ambiguity error the first `Any` is not bold?

**Comment by thchr** [comment]
That is nice! 
I do still think, subjectively, that the module names are the most visibly distinct feature in the printout, but it's a lot easier to disambiguate now (and definitely a ton clearer than the current styling).

I wonder if the module name coloring is helpful at all in this context though. To me, right now, they remain the most visible feature in the list, drawing my eye to them as a kind of semi-noisy "headers" of sorts (and notably, the headers come *after* the content - this is confuses me in the stack trace printing as well a bit; I'm always finding a "header", then I need to remind myself to look *upwards* to see its associated content).

If the modules were all colored gray, the "where-is-this-method-actually-from" line would serve as a nice and consistent "quasi-whitespace" of sorts between method signatures.
Just wondering out loud here, obviously - the proposed styling is already very nice..!

**Comment by mcabbott** [comment]
> Any idea why in the ambiguity error the first Any is not bold?

Probably a bad decision! In looking at examples with a few types & many `::Any`s, I was trying to reduce their importance. 

> I wonder if the module name coloring is helpful at all in this context though.

The hope is that it's a proxy for the package for whose types a method is being added. How often things will end up monochrome or looking like a rainbow in actual use, will depend... I guess colours will be most useful when there are just a few "foreign" methods, which are either the ones you care about, or the ones you can ignore.

> and notably, the headers come after the content - this is confuses me in the stack trace printing as well a bit

Agreed. There were other proposals in the giant stack trace threads, like highlighting the `[10]` or the function name, but (IIRC) these felt like visual puzzles to see what the colours went with. Qualifying the function names didn't look great, and moved further from how `@warn` etc. print locations.

Here, I worry that highlighting `Type{GitCredentialHelper}` would similarly introduce an annoying puzzle to solve. Whereas everyone is used to reading the stack traces (or soon will be!)

**Comment by JeffBezanson** [comment]
> ideal to share the common printing logic aside from those details.

This!

**Comment by mcabbott** [comment]
It would also be nice to unify the code which prints "MethodError: no method matching" with everything else --this PR doesn't change that, but affects "MethodError: ... ambiguous" because that is shared with `methods(f)`. But (IMO) further unification should be another PR.

For this one, I guess we need to decide whether to do this at all. And whether the style here is what's wanted? 
* Does the module colouring make sense? The module which owns the function gets no colour (excepts its sub-modules are blue), while "foreign" modules extending it are given colours like the stack traces (but starting cool instead of warm).
* Do bold types look OK on most people's screens?
* "MethodError: ... ambiguous" uses an option to print the module & line number on the same line as the method. This is not used in stack traces, should it be used in `methods(f)`?

* The length of the path & filename shown should be the same as for stack traces. This is quite long for e.g. LinearAlgebra: the error from `det([1 2 3; 4 5 6])` wraps at 132 characters for me. Perhaps we should revisit that.

I fixed some doctests but there are some more... I'd rather not do that too many times.

**Comment by simeonschaub** [comment]
This already looks great, sorry it got forgotten a bit! This seems preferable to #40916 since it is more consistent with stacktrace printing. 

>     * Does the module colouring make sense? The module which owns the function gets no colour (excepts its sub-modules are blue), while "foreign" modules extending it are given colours like the stack traces (but starting cool instead of warm).

Oh, that's smart! Sounds reasonable to me.

>     * Do bold types look OK on most people's screens?

Looks good for me, I wouldn't expect this to be a big issue.

>     * "MethodError: ... ambiguous" uses an option to print the module & line number on the same line as the method. This is not used in stack traces, should it be used in `methods(f)`?

Maybe? Although I don't think the more compact printing is bad here either.

>     * The length of the path & filename shown should be the same as for stack traces. This is quite long for e.g. LinearAlgebra: the error from `det([1 2 3; 4 5 6])` wraps at 132 characters for me. Perhaps we should revisit that.

Do you have a screenshot of how this looks like for you? This probably isn't a dealbreaker though.

> I fixed some doctests but there are some more... I'd rather not do that too many times.

Do you know about https://github.com/JuliaLang/julia/pull/40546#issuecomment-824380677? You can even pass `revise=true` to do this without having to rebuild the whole sysimage.

**Comment by mcabbott** [comment]
> Do you know about https://github.com/JuliaLang/julia/pull/40546#issuecomment-824380677? 

Now I do, thanks! Will try that out.

> Do you have a screenshot

Yes. This is the error printing with the pre-built Julia 1.6 -- some a long paths, and not useful ones. But perhaps orthogonal to this PR, which just re-uses the same path logic as for stack traces.

Screenshot 2021-05-25 at 10 14 14

File: doc/src/manual/methods.md [hunk 1]
```diff
@@ -211,26 +217,30 @@ of methods:
...
+     @ Dates /Users/me/.julia/dev/julia/usr/share/julia/stdlib/v1.7/Dates/src/periods.jl:372
+   [8] +(x::Rational, y::Integer)
+     @ rational.jl:310
+   [9] +(x::T, y::Integer) where {T<:AbstractChar}
+     @ char.jl:235
+  [10] +(Da::LinearAlgebra.Diagonal, Db::LinearAlgebra.Diagonal)
+     @ LinearAlgebra /Users/me/.julia/dev/julia/usr/share/julia/stdlib/v1.7/LinearAlgebra/src/diagonal.jl:172
 ...
-[180] +(a, b, c, xs...) in Base at operators.jl:424
+ [207] +(a::Any, b::Any, c::Any, xs::Any...)
```
Comment (simeonschaub) [inline doc/src/manual/methods.md:242]: Was this change intentional? I think always printing the `::Any` is a bit verbose and I am not sure it adds much information.

Comment (mcabbott) [reply doc/src/manual/methods.md:242]: I agree it's a bit verbose in cases like that. What it was meant to do was to help cases like this, where otherwise untyped arguments go missing:

Screenshot 2021-05-26 at 09 43 01

Comment (simeonschaub) [reply doc/src/manual/methods.md:242]: I think it would be fine to omit the `::Any` here, but I don't have any strong opinions on this.

Comment (mcabbott) [reply doc/src/manual/methods.md:242]: Here are some variants which print the module & path without a new line. (Because the new one above takes up a lot of lines compared to the old. And the types printed here are usually shorter than those in the stack traces.) I think this might be better. 

On the left I omit the `::Any`, what I don't like is that suddenly the variable name has to stand on its own, instead of being clearly secondary information to the signature. You can't answer "how many arguments?" by scanning "how many bold things?" anymore. 

On the right, I've also added bold to required keywords, shown with `cumsum`. Seems like useful information to mark them somehow, hope bold isn't confusing --- all the types are also requirements, after all.

Screenshot 2021-05-26 at 12 01 35

**Comment by tecosaur** [comment]
I made #40916 before someone pointed this out to me. I have no idea what's going to happen, but should this be the approach that ends up being used, I'd like to add some thoughts:
1. I'm not a fan of the two line version. If I'm looking at something with ~200 methods (like `+`) I'd rather have 200 lines to look at than 400. I think an "intelligent" switch could make sense though (if num methods < val, etc.)
2. I think argument names deserve more emphasis, in my PR I use the magenta colour for them. for example there's the `join` method with signature `join(strings, delim, last)` and I think it's worth applying emphasising argument names for cases such as those. If nothing else, since names are usually much shorter than type signatures this also makes it easier to count arguments, and resolves the `::Any` problem.
3. I really like how modules are given per-module colouring
4. This may be outside the scope of this PR, but I think it's worth considering applying [terminal hyperlinks](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda) to file names printed (in the terminal) by Julia

**Comment by mcabbott** [comment]
1. I think I now agree that one line is better, here. The types are typically much shorter than in stack traces. And moving the coloured module names to the right doesn't seem weird, especially since only a subset of them are coloured. (It's a trivial change but not pushed yet... https://github.com/JuliaLang/julia/pull/40546#issuecomment-824380677 leads to a segfault so I am updating things manually.)

2. The signature which identifies the method *is* these types, while the variable names are just some details of the source code. It's good to show both but I do think the emphasis should be on the types. If you are looking at `methods(f)`, I think it's often to see precisely what will match (or why it won't, or whether it will create ambiguity) rather than for general help.

3. That said, there's room for more exploration of syntax highlighting in general. I think OhMyREPL only touches things you type; but perhaps there should be hooks to make it easier to also change how errors (and the printing of `Expr`) etc. are printed? 

4. Something like this hyperlinking would be great, here and in stack traces. Especially if it meant you could not print the whole path, which is adding a lot of clutter to all these examples. But I think this should be an independent PR.

File: base/methodshow.jl [hunk 2]
```diff
@@ -206,24 +202,41 @@ function show(io::IO, m::Method)
...
     kwargs = kwarg_decl(m)
     if !isempty(kwargs)
         print(io, "; ")
-        join(io, map(sym_to_string, kwargs), ", ", ", ")
+        for kw in kwargs
+            skw = sym_to_string(kw)
+            if QuoteNode(kw) in m.roots # then it's required
+                printstyled(io, skw, color=:bold)
+            else
+                print(io, skw)
```
Comment (mcabbott) [inline base/methodshow.jl:225]: This is meant to check which keyword arguments are required, and print those in bold. 

Will this be robust? I don't know any details of the method struct, but in cases I could think of this appears to work.

Comment (mcabbott) [reply base/methodshow.jl:225]: Picture of this, BTW. Left is master, right this PR. 

Screenshot 2021-05-27 at 00 20 00

**Comment by tecosaur** [comment]
1. Cool
2. I'm not sure my thinking lights up with yours. If we were in black & white you could only emphasise one thing, but by using a colour you can brightly label multiple, and the eye can more easily jump to what you're looking for. See my PR for an example.
3. So do I, personally think it would be great if something like what I mock up in https://github.com/JuliaLang/julia/pull/40916#issuecomment-847497970 could happen, and OhMyREPL was included in Julia
4. I think this would make sense as an independent PR, yes. Affects this though, so I'll mention it here. If someone is willing to help me out (answer questions like which file should I put it in, how do I find X, ...) I'd like to do this as a learning experience for this codebase, since I like the idea of contributing to Julia :slightly_smiling_face: in particular I love how Julia has embraced Unicode, but I think it could benefit from going a bit further embracing colour and other features (like this)

**Comment by mcabbott** [comment]
Yes, it's true we aren't constrained to greyscale. I'm hesitant to touch the error printing, though!

I think the code which is printing the path is in errorshow.jl, this moves it into a function `print_module_path_file` but leaves it there. The path gets altered by `fixup_stdlib_path` in methodshow.jl. It's all pretty messy, but with `using Revise; Revise.track(Base)` you can fairly easily mess with it and see what breaks.

File: doc/src/manual/methods.md [hunk 3]
```diff
@@ -211,26 +211,21 @@ of methods:
...
-[16] +(x::BigInt, y::BigInt) in Base.GMP at gmp.jl:361
-[17] +(x::BigInt, c::Union{UInt16, UInt32, UInt64, UInt8}) in Base.GMP at gmp.jl:398
+# 207 methods for generic function "+" from Base:
+   [1] +(x::T, y::T) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} @ int.jl:87
+   [2] +(c::Union{UInt16, UInt32, UInt64, UInt8}, x::BigInt) @ Base.GMP gmp.jl:529
+   [3] +(c::Union{Int16, Int32, Int64, Int8}, x::BigInt) @ Base.GMP gmp.jl:535
+   [4] +(c::Union{UInt16, UInt32, UInt64, UInt8}, x::BigFloat) @ Base.MPFR mpfr.jl:376
+   [5] +(c::Union{Int16, Int32, Int64, Int8}, x::BigFloat) @ Base.MPFR mpfr.jl:384
+   [6] +(c::Union{Float16, Float32, Float64}, x::BigFloat) @ Base.MPFR mpfr.jl:392
+   [7] +(x::Union{Dates.CompoundPeriod, Dates.Period}) @ Dates /Users/me/.julia/dev/julia/usr/share/julia/stdlib/v1.7/Dates/src/periods.jl:372
```
Comment (mcabbott) [inline doc/src/manual/methods.md:221]: Not sure what path should be shown here; this one is obviously from me building locally. (It's not a doctest.)

**Comment by tecosaur** [comment]
Two things:
- Is anything more likely to happen with this in the near future?
- I'm liking @mcabbott's variables names, could this also make its way into stack traces perhaps?

**Comment by mcabbott** [comment]
It would be good to revive this, I think.

When last I looked there were a few doctest failures which I couldn't fix automatically for some reason.

Merge conflict is with #41929 which needs some care.

File: doc/src/manual/methods.md (hunk 3 above)
Comment (mcabbott) [reply doc/src/manual/methods.md:221]: On 1.7, these also don't look like the old docstring -- instead they have a path into the downloaded version, like `/Applications/Julia-1.7.app/.../LinearAlgebra/src/`:
```
julia> methods(+)
# 208 methods for generic function "+":
[1] +(x::T, y::T) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} in Base at int.jl:87
[2] +(c::Union{UInt16, UInt32, UInt64, UInt8}, x::BigInt) in Base.GMP at gmp.jl:529
[3] +(c::Union{Int16, Int32, Int64, Int8}, x::BigInt) in Base.GMP at gmp.jl:535
[4] +(c::Union{UInt16, UInt32, UInt64, UInt8}, x::BigFloat) in Base.MPFR at mpfr.jl:390
[5] +(c::Union{Int16, Int32, Int64, Int8}, x::BigFloat) in Base.MPFR at mpfr.jl:398
[6] +(c::Union{Float16, Float32, Float64}, x::BigFloat) in Base.MPFR at mpfr.jl:406
[7] +(x::Union{Dates.CompoundPeriod, Dates.Period}) in Dates at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Dates/src/periods.jl:372
[8] +(A::UpperHessenberg, B::UpperHessenberg) in LinearAlgebra at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/LinearAlgebra/src/hessenberg.jl:99
```

**Review by mcabbott — COMMENTED** [review]
Tests at last pass. The final appearance is like this (1.7 on the left):

Screenshot 2022-01-02 at 09 56 43

As noted, these changes also affect how methods are shown after an ambiguity error, but not after a methoderror. Ideally the latter would be updated to match, but not in this PR:

Screenshot 2022-01-02 at 10 03 09

The code used for printing the signature (with bold outer types, module colours) is the same as that used for printing stack traces (with light inner types, and module colours), except that it does not put the module & path on a new line. 

A few other comments/questions for review are highlighted below.

**Comment by tecosaur** [comment]
Great to see progress being made on this! @mcabbott I'm still keen on the idea of colouring variables names, in some cases I find it quite helpful — what are your thoughts?

**Comment by mcabbott** [comment]
I think it looks quite nice (here: https://github.com/JuliaLang/julia/pull/40251#issuecomment-849680473) but I'm not sure everyone will agree. And I think that it would be weird to do that here and not in stack traces. So this PR leaves that question for later.

**Comment by mcabbott** [comment]
Bump, pre-1.8. Pinging @StefanKarpinski maybe?

File: base/methodshow.jl (hunk 2 above)
Comment (vtjnash) [reply base/methodshow.jl:225]: no, this is not robust. but quite amusing that you found this

File: base/methodshow.jl [hunk 4]
```diff
@@ -206,24 +202,41 @@ function show(io::IO, m::Method)
...
+        end
     end
     print(io, ")")
     show_method_params(io, tv)
-    print(io, " in ", m.module)
-    if line > 0
-        file, line = updated_methodloc(m)
-        print(io, " at ", file, ":", line)
+
+    # module & flie, re-using function from errorshow.jl
```
Comment (vtjnash) [inline base/methodshow.jl:235]:
```suggestion
    # module & file, re-using function from errorshow.jl
```

File: base/methodshow.jl [hunk 5]
```diff
@@ -266,12 +284,31 @@ function show_method_table(io::IO, ms::MethodList, max::Int=-1, header::Bool=tru
...
     last_shown_line_infos === nothing || empty!(last_shown_line_infos)
 
+    modul = if mt === _TYPE_NAME.mt  # type constructor
+            which(ms.ms[1].module, ms.ms[1].name)
+        else
+            mt.module
+        end
+    modulecolordict = Dict{Module, Symbol}()
+    modulecolorcycler = Iterators.Stateful(Iterators.cycle(METHODLIST_MODULECOLORS))
+    modulecolordict[parentmodule_before_main(modul)] = :blue
```
Comment (vtjnash) [inline base/methodshow.jl:294]: Is this really our only strategy for this? Global state through a particular list?

File: doc/src/devdocs/inference.md [hunk 6]
```diff
@@ -96,7 +96,7 @@ Each statement gets analyzed for its total cost in a function called
 as follows:
 ```jldoctest; filter=r"tuple.jl:\d+"
 julia> Base.print_statement_costs(stdout, map, (typeof(sqrt), Tuple{Int},)) # map(sqrt, (2,))
-map(f, t::Tuple{Any}) in Base at tuple.jl:179
+map(f::Any, t::Tuple{Any}) @ Base tuple.jl:218
```
Comment (vtjnash) [inline doc/src/devdocs/inference.md:99]: this seems like a regression

**Review by vtjnash — APPROVED** [review]
It looks like this seems sound. I have mentioned a few places to fix (to not change), but otherwise seems like it successfully shares code, so it quite sane.

File: doc/src/devdocs/inference.md (hunk 6 above)
Comment (mcabbott) [reply doc/src/devdocs/inference.md:99]: Can you explain what you mean? 

I just landed here by chasing errors, and have not looked for the code which generates this. I believe that `Base.Math.throw_complex_domainerror(:sqrt::Symbol, %2::Float64)::Union{}` just below in this output does not take the path this PR changes, since it does not get printed in bold:

Screenshot 2022-01-19 at 18 09 22

Is there an alternative path that the first line's printing should take?

File: base/methodshow.jl (hunk 5 above)
Comment (mcabbott) [reply base/methodshow.jl:294]: I have no memory of writing this :/  I presume I tried to copy the stack traces.

The list [:cyan, :green, :yellow] is global, but isn't `modulecolorcycler` local to this function?

File: base/methodshow.jl (hunk 2 above)
Comment (mcabbott) [reply base/methodshow.jl:225]: Ok. I just poke around... Is there a robust way to query this?

Comment (vtjnash) [reply base/methodshow.jl:225]: I don't know of one right now. Let's leave that for later work?

File: doc/src/devdocs/inference.md (hunk 6 above)
Comment (vtjnash) [reply doc/src/devdocs/inference.md:99]: The added `::Any` is unnecessary, and seems to be explicitly being added in this PR

File: base/methodshow.jl (hunk 5 above)
Comment (vtjnash) [reply base/methodshow.jl:294]: ah, true, seems to be a copy from there. I guess this is okay

File: doc/src/devdocs/inference.md (hunk 6 above)
Comment (mcabbott) [reply doc/src/devdocs/inference.md:99]: Ok, I can surely figure out a way to remove it here. 

But the logic of including it in general is this:

https://user-images.githubusercontent.com/32575566/119697648-6b188f00-be1e-11eb-91e3-c7ae8612f2e4.png

On the right you can count the arguments by counting the bold words. On the left the grey `f,` is really easy to miss, and seems worse than before.

**Comment by JeffBezanson** [comment]
From triage: after much discussion, we all came to a different conclusion, that we like having the signature and location on two lines since (1) it's nice to be consistent with stack traces, and (2) the method location is actually always quite important and two lines makes it much easier to scan the locations. We also like omitting `::Any`; it's less noisy and matches how we expect people to write such method signatures in their code.

Sorry about all the back-and-forth on this; these formatting PRs can be really contentious as you know :)

**Comment by mcabbott** [comment]
Oh I know. 

One question though is whether the all-important `f` fist argument goes missing too easily, light grey and no type. That's what the `::Any` was trying to address. Another way might be to make all the arguments hot pink, a bit like https://github.com/JuliaLang/julia/pull/40251#issuecomment-849680473 . Strong feelings on this? 

Fixing the code is easy, fixing the tests may take me a while to get to, we'll see.

**Comment by vtjnash** [comment]
Could we add some logic that prints either `<light>::<bold>` or `<normal>`, depending on whether there is content to print on the right-hand side?

**Comment by mcabbott** [comment]
Yes, surely. Will try that out.

**Comment by tecosaur** [comment]
Jeff/Jameson as a fan of making the arg name stand out regardless (as in https://github.com/JuliaLang/julia/pull/40251#issuecomment-849680473), is there anything you have against that?

**Comment by vtjnash** [comment]
I mildly dislike the extra color, on the basis of being too noisy.

**Comment by vtjnash** [comment]
Note: closes #40913

**Comment by tecosaur** [comment]
> I mildly dislike the extra color, on the basis of being too noisy.

I see. I'd comment that having it be styled the same regardless is good for visual consistency and it isn't unimportant enough to be greyed out, as the naming can often be informative.

**Comment by JeffBezanson** [comment]
Ah, I forgot to mention that @staticfloat made the point that we shouldn't rely too heavily on colors; the output should be nice even with the colors stripped. That's another reason I prefer leaving out `::Any`.
